### PR TITLE
docs: add contribution guidelines

### PR DESCRIPTION
**Pull request by mojomex**
## PR Type

- Improvement

## Related Links

- Inspired by [comma.ai's openpilot](https://github.com/commaai/openpilot/blob/master/docs/CONTRIBUTING.md)

## Description

Add contribution guidelines to Nebula. This allows easier contributions from the community and provides a clear set of rules and expectations we can refer to.

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: docs/contribute.md [hunk 1]
```diff
@@ -1,3 +1,87 @@
...
 
-WIP - please check back soon!
+Thank you for your interest in contributing to Nebula!
+We’re excited to work with the open-source community to build an even better project.
+Before submitting your contributions, please take a moment to review our guidelines to ensure a smooth process.
+
+## Getting Started
+
+- Set up your development environment: [installation](installation.md)
+- Enable [pre-commit](https://pre-commit.com/#install)
```
Comment (drwnz) [inline docs/contribute.md:10]: Optional suggestion: open external links in a new tab.
```suggestion
- Enable <a href="https://pre-commit.com/#install" target="_blank">pre-commit</a>
```

File: docs/contribute.md [hunk 2]
```diff
@@ -1,3 +1,87 @@
...
+  # Install pre-commit on your system
+  pip3 install pre-commit
+
+  # Enable pre-commit for Nebula
+  cd ~/nebula_ws/src
+  pre-commit install
+  ```
+
+- Set up your IDE to use Clang-Tidy with the `.clang-tidy` configuration found in the repository root
+- Docs are found at [tier4.github.io/nebula](https://tier4.github.io/nebula)
```
Comment (drwnz) [inline docs/contribute.md:22]:
```suggestion
- Documentation can be found at [tier4.github.io/nebula](https://tier4.github.io/nebula)
```

File: docs/contribute.md [hunk 3]
```diff
@@ -1,3 +1,87 @@
...
+
+## How to Contribute
+
+### About Large Contributions
```
Comment (drwnz) [inline docs/contribute.md:26]:
```suggestion
### About large contributions
```

File: docs/contribute.md [hunk 4]
```diff
@@ -1,3 +1,87 @@
...
+This prevents double work from other community members.
+
+Do not be afraid to open a draft PR in an early stage of your implementation.
+This gives us a chance to spot potential problems early (e.g. line count) and lets us align with you better.
+
+### What gets merged?
+
+We are keen on merging PRs that are clear wins for the project:
+
+- support for popular sensor models
```
Comment (drwnz) [inline docs/contribute.md:43]:
```suggestion
- Support for popular sensor models
```

File: docs/contribute.md [hunk 5]
```diff
@@ -1,3 +1,87 @@
...
+- unit tests
```
Comment (drwnz) [inline docs/contribute.md:44]:
```suggestion
- Unit tests
```

File: docs/contribute.md [hunk 6]
```diff
@@ -1,3 +1,87 @@
...
+- bug fixes
```
Comment (drwnz) [inline docs/contribute.md:45]:
```suggestion
- Bug fixes
```

File: docs/contribute.md [hunk 7]
```diff
@@ -1,3 +1,87 @@
...
+- robustness improvements, error handling and reporting
```
Comment (drwnz) [inline docs/contribute.md:46]:
```suggestion
- Improvements for robustness, error handling, and reporting
```

File: docs/contribute.md [hunk 8]
```diff
@@ -1,3 +1,87 @@
...
+If a PR has value for the project but takes a long time or a lot of effort to review,
+we will have to reject it. We will let you know if we think we can handle the scope of your
+contribution in the discussion of your issue.
+
+### What doesn't get merged?
+
+- 500+ line PRs: split your PR into smaller ones
+- PRs without sufficient unit tests: if unit tests are not rigorous enough or have low coverage
+- PRs with multiple unrelated changes: split up into multiple PRs
+- performance-heavy features: features with latency increases of more than a low single-digit percentage
```
Comment (drwnz) [inline docs/contribute.md:57]:
```suggestion
- Performance-heavy features: features with latency increases of more than a low single-digit percentage
```

File: docs/contribute.md [hunk 9]
```diff
@@ -1,3 +1,87 @@
...
+- style changes
```
Comment (drwnz) [inline docs/contribute.md:58]:
```suggestion
- Style changes
```

File: docs/contribute.md [hunk 10]
```diff
@@ -1,3 +1,87 @@
...
+- big refactorings
```
Comment (drwnz) [inline docs/contribute.md:59]:
```suggestion
- Big refactorings
```

File: docs/contribute.md [hunk 11]
```diff
@@ -1,3 +1,87 @@
...
+- features implemented only for some vendors
```
Comment (drwnz) [inline docs/contribute.md:60]:
```suggestion
- Features implemented only for some vendors
```

File: docs/contribute.md [hunk 12]
```diff
@@ -1,3 +1,87 @@
...
+
+## Pull Requests
+
+Pull requests should be against the `develop` branch.
+Hotfix PRs should be against the `master` branch.
```
Comment (drwnz) [inline docs/contribute.md:65]: This one is important!

```suggestion
Hotfix PRs should be against the `main` branch.
```

File: docs/contribute.md [hunk 13]
```diff
@@ -1,3 +1,87 @@
...
+
+A good PR fulfills all of the following:
+
+- passes pre-commit
```
Comment (drwnz) [inline docs/contribute.md:69]:
```suggestion
- Passes pre-commit
```

File: docs/contribute.md [hunk 14]
```diff
@@ -1,3 +1,87 @@
...
+- conforms to the repo's `.clang-tidy` and `.clang-format`
```
Comment (drwnz) [inline docs/contribute.md:70]:
```suggestion
- Conforms to the repo's `.clang-tidy` and `.clang-format`
```

File: docs/contribute.md [hunk 15]
```diff
@@ -1,3 +1,87 @@
...
+- passes all CI checks
```
Comment (drwnz) [inline docs/contribute.md:71]:
```suggestion
- Passes all CI checks
```

File: docs/contribute.md [hunk 16]
```diff
@@ -1,3 +1,87 @@
...
+- a single clearly stated goal
```
Comment (drwnz) [inline docs/contribute.md:72]:
```suggestion
- Has a single clearly stated goal
```

File: docs/contribute.md [hunk 17]
```diff
@@ -1,3 +1,87 @@
...
+- every line changed directly contributes to that goal
```
Comment (drwnz) [inline docs/contribute.md:73]:
```suggestion
- Every line changed directly contributes to that goal
```

File: docs/contribute.md [hunk 18]
```diff
@@ -1,3 +1,87 @@
...
+- unit tests for added/fixed functionality where applicable
```
Comment (drwnz) [inline docs/contribute.md:74]:
```suggestion
- Includes unit tests for added/fixed functionality where applicable
```

File: docs/contribute.md [hunk 19]
```diff
@@ -1,3 +1,87 @@
...
+- documentation for added functionality
```
Comment (drwnz) [inline docs/contribute.md:75]:
```suggestion
- New functionality is documented
```

File: docs/contribute.md [hunk 20]
```diff
@@ -1,3 +1,87 @@
...
+- a reproducible evaluation / testing method where applicable
```
Comment (drwnz) [inline docs/contribute.md:76]:
```suggestion
- Has a reproducible evaluation/testing method where applicable
```

File: docs/contribute.md [hunk 21]
```diff
@@ -1,3 +1,87 @@
...
+  - e.g. Rosbags or PCAPs attached
```
Comment (drwnz) [inline docs/contribute.md:77]:
```suggestion
    - e.g. Rosbags or PCAPs attached
```

File: docs/contribute.md [hunk 22]
```diff
@@ -1,3 +1,87 @@
...
+  - for performance optimizations, easily runnable benchmarks
```
Comment (drwnz) [inline docs/contribute.md:78]:
```suggestion
  - Easily runnable benchmarks are included for performance optimizations
```

File: docs/contribute.md [hunk 23]
```diff
@@ -1,3 +1,87 @@
...
+
+We cannot start a review before pre-commit and CI are passing.
+
+## Other Ways to Contribute
+
+Bug reports in GitHub issues:
+
+- provide logs, PCAPs and videos where possible
```
Comment (drwnz) [inline docs/contribute.md:86]:
```suggestion
- Please provide logs, PCAPs and videos where possible
```

File: docs/contribute.md [hunk 24]
```diff
@@ -1,3 +1,87 @@
...
+- specify the exact version and environment used
```
Comment (drwnz) [inline docs/contribute.md:87]:
```suggestion
- Make sure you always specify the exact version and environment used
```

**Review by drwnz — CHANGES_REQUESTED** [review]
Mostly minor (some more style-based that could go either way), with one or two must-fixes in there, but generally excellent.

File: docs/contribute.md [hunk 25]
```diff
@@ -1,3 +1,87 @@
...
+
+- Passes pre-commit
+- Conforms to the repo's `.clang-tidy` and `.clang-format`
+- Passes all CI checks
+- Has a single clearly stated goal
+- Every line changed directly contributes to that goal
+- Includes unit tests for added/fixed functionality where applicable
+- New functionality is documented
+- Has a reproducible evaluation/testing method where applicable
+    - e.g. Rosbags or PCAPs attached
```
Comment (mojomex) [inline docs/contribute.md]:
```suggestion
  - e.g. Rosbags or PCAPs attached
```

**Comment by mojomex** [comment]
@drwnz Thanks for the many suggestions! I applied them all. Please re-review when you find the time :pray:

**Review by drwnz — APPROVED** [review]
Nice ;)

**Pull request merged**